android/src/main/kotlin/com/example/user_location_plugin/UserLocationPlugin.kt

```diff
@@ -1,35 +1,141 @@
 import io.flutter.plugin.common.MethodChannel.Result
+import io.flutter.plugin.common.PluginRegistry
 
-/** UserLocationPlugin */
-class UserLocationPlugin: FlutterPlugin, MethodCallHandler {
-  /// The MethodChannel that will the communication between Flutter and native Android
-  ///
-  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
-  /// when the Flutter Engine is detached from the Activity
-  private lateinit var channel : MethodChannel
-
-  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
-    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "user_location_plugin")
-    channel.setMethodCallHandler(this)
-  }
-
-  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
-    if (call.method == "getPlatformVersion") {
-      result.success("Android ${android.os.Build.VERSION.RELEASE}")
-    } else {
-      result.notImplemented()
-    }
-  }
-
-  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
-    channel.setMethodCallHandler(null)
-  }
+
+class UserLocationPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
+    PluginRegistry.ActivityResultListener {
+    private lateinit var channel: MethodChannel
+    private lateinit var context: Context
+    private lateinit var activity: Activity
+    private lateinit var permissionEventChannel: EventChannel
+    private lateinit var permissionEventSource: EventChannel.EventSink
+    private var permissionStreamHandler: EventChannel.StreamHandler =
+        object : EventChannel.StreamHandler {
+            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
+                if (events != null) {
+                    permissionEventSource = events
+                }
+            }
+
+            override fun onCancel(arguments: Any?) {
+            }
+        }
+
+    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+        context = flutterPluginBinding.applicationContext
+        channel = MethodChannel(flutterPluginBinding.binaryMessenger, USER_LOCATION_PLUGIN)
+        channel.setMethodCallHandler(this)
+        permissionEventChannel =
+            EventChannel(flutterPluginBinding.binaryMessenger, PERMISSION_EVENT_CHANNEL)
+        permissionEventChannel.setStreamHandler(permissionStreamHandler)
+    }
+
+    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
+        when (call.method) {
+            GET_PLATFORM_VERSION -> getPlatformVersion(result)
+            REQUEST_PERMISSION -> requestPermission(result)
+            CHECK_PERMISSION -> checkPermission(result)
+            else -> result.notImplemented()
+        }
+    }
+
+    private fun getPlatformVersion(result: Result) {
+        result.success("Android ${android.os.Build.VERSION.RELEASE}")
```

Comment:
You can add the Android String into the companion object

Reply:
added